### PR TITLE
WICKET-6427 Fire an event once all ajax timers are registered

### PR DESCRIPTION
**Pull request by Jezza**
Fire Wicket.Event.Topic.AJAX_TIMERS_BOUND event once all Wicket.Timer.set() calls are executed

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java [hunk 1]
```diff
@@ -353,6 +353,7 @@ private void renderCombinedEventScripts()
 		}
 		if (combinedScript.length() > 0)
 		{
+			combinedScript.append("\nWicket.Event.publish(Wicket.Event.Topic.AJAX_TIMERS_BOUND);");
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: How do you know that any timers are in the response ?
It looks to me that this event is published no matter whether there are timer behaviors in use or not.

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: It follows the same logic as the [AJAX_HANDLERS_BOUND](https://github.com/apache/wicket/blob/master/wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java#L339) event.

Which is always fired if there's `OnDomReadyHeaderItem`s.

If the event name is confusing, then I'd recommend changing both of them to suit their actual function.

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: The events themselves are closer to post-`load`s or post-`domready`s

Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: Right!
The names are confusing!
They should use `load` and `domready` in their names to be more clear.
What about `Wicket.Event.Topic.ONDOMREADY_EVENTS_BOUND` and `...ONLOAD_EVENTS_BOUND` ?

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: I'd be happy to change it to that.
Should we offer some level of backwards compat, and alias `AJAX_HANDLERS_BOUND`?

Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: Yes, I think we should publish both for 8.x and remove `AJAX_HANDLERS_BOUND` for 9.x.
We can also log a warning when `AJAX_HANDLERS_BOUND` published to a subscriber.

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: Also, I noticed that it doesn't work for header items added via the `AjaxRequestTarget`.
Should that be a concern?

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: That actually might just be my code specifically. I'll have to take a look tomorrow.

Comment (Jezza) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: There we go, I pushed the latest state of the branch.

File: wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java [hunk 2]
```diff
@@ -336,7 +336,7 @@ private void renderCombinedEventScripts()
 		}
 		if (combinedScript.length() > 0)
 		{
-			combinedScript.append("\nWicket.Event.publish(Wicket.Event.Topic.AJAX_HANDLERS_BOUND);");
+			combinedScript.append("\nWicket.Event.publish(Wicket.Event.Topic.AJAX_ONDOMREADY_EVENTS_BOUND);");
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: I think we should keep both here, to be backward compatible.

File: wicket-core/src/main/java/org/apache/wicket/ajax/res/js/wicket-event-jquery.js [hunk 3]
```diff
@@ -277,6 +277,10 @@
 			* @param topic {String} - the channel name for which all subscribers will be notified.
 			*/
 			publish: function (topic) {
+				if (topic === Topic.AJAX_HANDLERS_BOUND) {
+            		console.warn("As of Wicket 8, Topic.AJAX_HANDLERS_BOUND has been deprecated in favour of Topic.AJAX_ONDOMREADY_EVENTS_BOUND and will be removed in Wicket 9.")
+            		topic = Topic.AJAX_ONDOMREADY_EVENTS_BOUND;
+            	}
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/ajax/res/js/wicket-event-jquery.js]: To be backward compatible we will have to still publish this event together with the new one.
This check and warning should be moved to `#subscribe()`.

**Comment by Jezza** [comment]
Pushed the latest changes.

Though that does spring a question to mind, the events are only published if there's `OnDomReady` or `OnLoad` events, so should we just always include them?

**Comment by svenmeier** [comment]
>should we just always include them

Good question. IMHO it would make sense to always fire these 'new' events.

**Comment by Jezza** [comment]
Alrighty, preliminary check with that in place:
It breaks 214 tests, so it's gonna take a bit to update all of the tests.

**Comment by Jezza** [comment]
I've pushed the state so you can take a look as I'm working on updating the tests, so I wouldn't recommend accepting the PR just yet.

**Comment by Jezza** [comment]
Should I take the opportunity to update some of the tests, such as `AttributeModifierComponentTest`.
It seems to use an old way of verifying rendered content.
I could migrate it over to the, seemingly, standardised `_expected.html` variant.

**Comment by Jezza** [comment]
I'm moving my branches around, so I guess github spotted that it's no longer on my master branch.

I, personally, don't have any need of this ticket anymore, and honestly, there's a lot of work to do.
I think I got, maybe, 25% of the way through the tests.
I did have to change all of them, so I think unless someone wants it, I'll just let it lie.

**Pull request closed without merging**